CppImport/src/macro/LexicalHelper.cpp

```diff
@@ -0,0 +1,283 @@
+	else
+		start = range.getBegin();
+
+	if (isConcatenationStringifycation(range.getEnd()))
+		end = clang_->sourceManager()->getImmediateExpansionRange(range.getEnd()).second;
+	else
+		end = range.getEnd();
+
+	return clang::SourceRange(start, end);
+}
+
+QString LexicalHelper::unexpandedSpelling(clang::SourceRange range)
+{
+	auto result = clang_->spelling(unexpandedSourceRange(range));
+	while (result.startsWith("\\")) result = result.right(result.length() - 1);
+
+	return result.trimmed();
+}
+
+void LexicalHelper::correctNode(clang::Decl* clangAstNode, Model::Node* envisionAstNode)
+{
+	clang::SourceRange spellingRange = clangAstNode->getSourceRange();
+
+	if (auto fDecl = clang::dyn_cast<clang::FunctionDecl>(clangAstNode))
+	{
+		spellingRange = fDecl->getNameInfo().getSourceRange();
+
+		auto ooMethod = DCast<OOModel::Method>(envisionAstNode);
+		Q_ASSERT(ooMethod);
+
+		if (ooMethod->results()->size() > 0)
+			correctNode(fDecl->getReturnTypeSourceRange(), ooMethod->results()->first());
+
+		int i = 0;
+		for (auto it = fDecl->param_begin(); it != fDecl->param_end(); it++)
+		{
+			auto varDecl = clang::dyn_cast<clang::VarDecl>(*it);
+			Q_ASSERT(varDecl);
+
+			auto start = varDecl->getLocation();
+			auto end = varDecl->hasInit() ? varDecl->getInit()->getSourceRange().getBegin() :
+													  varDecl->getSourceRange().getEnd();
+
+			auto ooVarDecl = DCast<OOModel::VariableDeclaration>(ooMethod->arguments()->at(i));
+
+			correctNode(varDecl->getTypeSourceInfo()->getTypeLoc().getSourceRange(),
+												ooVarDecl->typeExpression());
+			correctNode(clang::SourceRange(start, end), ooVarDecl);
+			i++;
+		}
+	}
+	else if (auto varDecl = clang::dyn_cast<clang::VarDecl>(clangAstNode))
+	{
+		auto ooVarDecl = DCast<OOModel::VariableDeclaration>(envisionAstNode);
+		Q_ASSERT(ooVarDecl);
+
+		auto start = varDecl->getLocation();
+		auto end = varDecl->getSourceRange().getEnd();
+
+		correctNode(varDecl->getTypeSourceInfo()->getTypeLoc().getSourceRange(),
+											ooVarDecl->typeExpression());
+
+		spellingRange = clang::SourceRange(start, end);
+	}
+	else if (auto namedDecl = clang::dyn_cast<clang::NamedDecl>(clangAstNode))
+	{
+		spellingRange = clang::SourceRange(namedDecl->getLocation(), namedDecl->getSourceRange().getEnd());
+	}
+
+	correctNode(spellingRange, envisionAstNode);
+}
+
+void LexicalHelper::correctNode(clang::Stmt* clangAstNode, Model::Node* envisionAstNode)
+{
+	if (auto callExpr = clang::dyn_cast<clang::CallExpr>(clangAstNode))
+		correctNode(callExpr->getCallee()->getSourceRange(),
+											envisionAstNode);
+	else if (auto constrExpr = clang::dyn_cast<clang::CXXConstructExpr>(clangAstNode))
+	{
+		correctNode(
+				clang::SourceRange(constrExpr->getLocStart(),
+										 constrExpr->getParenOrBraceRange().getBegin()),
+											envisionAstNode);
+	}
+	else
+		correctNode(clangAstNode->getSourceRange(), envisionAstNode);
+}
+
+void LexicalHelper::correctNode(clang::SourceRange range, Model::Node* original)
```

Comment:
You need to explain somewhere, what all these correctXXX methods are and why they are necessary.